src/ptvsd/_vendored/pydevd/_pydevd_bundle/pydevd_resolver.py

```diff
@@ -212,24 +235,51 @@ def resolve(self, dict, key):
         raise UnableToResolveVariableException()
 
     def key_to_str(self, key):
-        if isinstance(key, str):
-            return '%r' % key
-        else:
-            if not pydevd_constants.IS_PY3K:
-                if isinstance(key, unicode):
-                    return "u'%s'" % key
-            return key
+        if not pydevd_constants.IS_PY3K:
+            if isinstance(key, unicode):
+                return "u'%s'" % key
+        return '%r' % key
 
     def init_dict(self):
         return {}
 
+    def get_contents_debug_adapter_protocol(self, dct):
+        '''
+        This method is to be used in the case where the variables are all saved by its id (and as
+        such don't need to have the `resolve` method called later on, so, keys don't need to
+        embed the reference in the key).
+
+        Note that the return should be ordered.
+
+        :return list(tuple(name:str, value:object, evaluateName:str))
+        '''
+        ret = []
+
+        i = 0
+        for key, val in dict_iter_items(dct):
+            i += 1
+            key_as_str = self.key_to_str(key)
+            ret.append((key_as_str, val, '[%s]' % (key_as_str,)))
+            if i > MAX_ITEMS_TO_HANDLE:
+                ret.append((TOO_LARGE_ATTR, TOO_LARGE_MSG))
+                break
+
+        ret.append(('__len__', len(dct), '.__len__()'))
```

Comment:
There's a disparity here in that the value it reports is `len()`, but it's presented as the result of calling `__len__()`. These can be different in some corner cases, since `len()` does a bunch of type conversions on what `__len__` returns.

In a similar case elsewhere, we used synthetic members that are presented as function calls - i.e. instead of `d.__len__(x)`, it would show up as a member `len(d)` in the IDE, and marked differently from attributes. Might be something to consider here.

Reply:
Agreed. Created https://github.com/Microsoft/ptvsd/issues/1176 to address that (ptvsd was giving `d.__len__`, before, so, it shouldn't be a blocker -- and I think the changes needed for that will be clearer in a new pull request).